### PR TITLE
Update taint-and-toleration.md

### PR DESCRIPTION
**Pull request by dongzhiwei-git**
When I was studying, I felt like I didn't understand something here. Note here, I think it is necessary and hope to adopt it

### CODE REVIEW

**Comment by windsonsea** [comment]
It would be better to include the reasons for making this change in the description section of this PR.

image

**Comment by dongzhiwei-git** [comment]
> It would be better to include the reasons for making this change in the description section of this PR.
> 
> image

Done

**Review by kundan2707 — COMMENTED** [review]
reason for changes is missing in description

**Comment by dongzhiwei-git** [comment]
> > It would be better to include the reasons for making this change in the description section of this PR.
> > image
> 
> Done


Thank you very much for your review, maybe I didn’t describe it very clearly. Maybe you can read my modifications and understand my description better.

**Comment by tengqm** [comment]
> Thank you very much for your review, maybe I didn’t describe it very clearly. Maybe you can read my modifications and understand my description better.

A contributor is responsible to describe a PR clearly for the reviewers to (quickly) check what it is about. This applies to the caption and the description of a PR. Both this message can be searched from git history.

You modifications are not so clear as you believed. That is another reason we are asking for an explanation. When `effects` is empty, the operator must be "Equal"? The value must equal to what?

**Comment by dongzhiwei-git** [comment]
> > Thank you very much for your review, maybe I didn’t describe it very clearly. Maybe you can read my modifications and understand my description better.
> 
> A contributor is responsible to describe a PR clearly for the reviewers to (quickly) check what it is about. This applies to the caption and the description of a PR. Both this message can be searched from git history.
> 
> You modifications are not so clear as you believed. That is another reason we are asking for an explanation. When `effects` is empty, the operator must be "Equal"? The value must equal to what?

When `operator` is "Equal"., The `value` is not empty

**Comment by utkarsh-singh1** [comment]
> When operator is "Equal"., The value is not empty

Yeah, I tested on condition with empty value with key and Equal effect, but it does have not any effect, so it is necessary to have a value, key and equal operator in case of empty effect, also according to this doc the value here that can be mentioned is `value1` .

**Comment by tengqm** [comment]
> > When operator is "Equal"., The value is not empty
> 
> Yeah, I tested on condition with empty value with key and Equal effect, but it does have not any effect, so it is necessary to have a value, key and equal operator in case of empty effect, also according to this doc the value here that can be mentioned is `value1` .

Sorry I still don't understand what you are saying here.  What do you mean by "it is necessary to have a value, key and equal operator in case of empty effect"?
We have to be precise here. An empty effect is an empty string or simply a field that is not specified? Suppose we mean an empty string here, we mean that "if the effect is an empty string, and we have `operator` set to `"Equal"`, we have to provide both `key` and `value`."?

When we set the `operator` to `"Equal"`, the `value` cannot be empty, this has nothing to do with `effect`, ... is this true?

I can feel that there is an ambiguity here to be clarified, but I don't think we are fixing it correctly in the PR as is.

**Comment by utkarsh-singh1** [comment]
@tengqm,

 I have tainted the node -
image



And This is the `pod-def.yaml` file I am using -  

```yaml
apiVersion: v1
kind: Pod
metadata:
    name: pod5
    labels:
        app: hello
spec:
    containers:
        - name: image5
          image: redis
    tolerations:
        - key: "app"
          value: "be" 
          effect: "" 
          operator: "Equal"
```

**Comment by utkarsh-singh1** [comment]
I have tested on emptied effect - 
1. without any empty string and 
2. with empty string, 

in Both cases pod scheduled on tainted node.

**Comment by tengqm** [comment]
> I have tested on emptied effect -
> 
>     1. without any empty string and
> 
>     2. with empty string,
> 
> 
> in Both cases pod scheduled on tainted node.

Thanks for the experiments.
What if `effect` is empty, and `operator` is `Exists`?

**Comment by utkarsh-singh1** [comment]
Here is both the result with -

1. Without an empty string in effect - 

Screenshot_20231020_141834

2. With an empty string in effect - 

Screenshot_20231020_141720

**Comment by tengqm** [comment]
@utkarsh-singh1 You got to understand this:

```
- key: something
  effect:
```
is equivalent to
```
- key: something
  effect: ""
```
What I meant to experiment is acutally this:
```
- key: something
  # ignore effect
```

**Comment by tengqm** [comment]
@utkarsh-singh1 The last experiment is whether we can have `operator` set to `Exists` and `effect` ignored. No need to paste the result.
With all these experiment results at hand, we then know how to revise the texts.

**Comment by utkarsh-singh1** [comment]
@tengqm, in case of an empty `effect` and `operator` is `Exists`-

1. If the `key` is not empty then the pod will not schedule on a tainted node that does not match the key , and schedule on the tainted node where the key matches.

2. If `key` is empty then pod will schedule on any tainted node , as it is already mentioned above.

**Comment by tengqm** [comment]
@utkarsh-singh1 Thank you very much for these verification.

The current saying "An empty `effect` matches all effects with key `key1`" is too brief, but it is correct. The matching has nothing to do with "operator" or "value" then. Correct?

**Comment by utkarsh-singh1** [comment]
@tengqm, 

> An empty effect matches all effects with key key1

1. ### This will be true if the node is tainted in this format - 

```bash
kubectl taint node node_name key1:taint-effect
```
Also,
- empty ⇒ `value`, `effect` and  `operator` but with `key` ⇾ `key1`, pod will schedule on tainted node.
- empty ⇒ `value`, `effect `, but `operator` ⇒ "Exists" or "Equal" and  `key` ⇒ `key1`, pod will schedule on tainted node.
- empty ⇒ `value`, `effect` and  `key` but with `operator` ⇾ `Exists`, pod will tolerate any tainted node.




2. ### If node is tainted in this format - 

```bash
kubectl taint node node_name key1=value1:taint-effect
```
 Also,

- empty ⇒ `effect`, `value` but `operator` ⇒ `Exists` and `key` ⇒ `key1`, pod schedules on tainted node.
- empty ⇒ `effect`, `value` but `operator` ⇒ `Equal` and `key` ⇒ `key1`, pod doesn't schedule on tainted node.
- empty ⇒ `effect` but `key` ⇒ `key1`, `value` ⇒ `value1` and `operator` ⇒ `Equal`, pod schedule on tainted node.
- empty ⇒ `value`, `effect` and  `operator` but with `key` ⇾ `key1`, pod will not schedule on tainted node.

All of this test done on being taint_effect being NoSchedule.

**Comment by tengqm** [comment]
@utkarsh-singh1 Many thanks to the summary. Based on this summary, do we have any suggestions to improve the current description? I'm not talking about a table that enumerates all kinds of combinations. I'm thinking of just one or two bullets.

**Comment by utkarsh-singh1** [comment]
@tengqm, This statement - 
> An empty `key` with operator `Exists` matches all keys, values and effects which means this will tolerate everything.

Seems to be working for both taint conditions.

If this statement -
> An empty `effect` matches all effects with key `key1`.

can be changes to - 
> An empty `effect` matches all affects with key `key1` and operator "Exists" will schdule on tainted node.

File: content/en/docs/concepts/scheduling-eviction/taint-and-toleration.md [hunk 1]
```diff
@@ -80,7 +80,7 @@ There are two special cases:
 An empty `key` with operator `Exists` matches all keys, values and effects which means this
 will tolerate everything.
 
-An empty `effect` matches all effects with key `key1`.
+An empty `effect` matches all effects with key `key1`(the `operator` is `Equal` and the `values` are equal).
```
Comment (sftim) [inline content/en/docs/concepts/scheduling-eviction/taint-and-toleration.md:83]: I don't think this is true. Leaving `effect` unset does not force the `operator` to be Equal.

Can you provide evidence that I've got it wrong? If not, I don't think we should merge the change.

**Review by sftim — CHANGES_REQUESTED** [review]

**Comment by divya-mohan0209** [comment]
@dongzhiwei-git : Please could you advise if you're able to address suggestions made by @sftim in [this review](https://github.com/kubernetes/website/pull/43572#discussion_r1501869863)?

**Comment by divya-mohan0209** [comment]
@dongzhiwei-git : There has been no response on this PR for ~2 months & due to lack of traction, we'll be closing it off. If you'd like to resume work on it, please issue the `/reopen` command.

/close

**Pull request closed without merging**